cpp/src/groupby/groupby.cu

```diff
@@ -193,23 +194,37 @@ detail::sort::sort_groupby_helper& groupby::helper()
   return *_helper;
 };
 
-std::pair<std::unique_ptr<table>, std::unique_ptr<column>> groupby::shift(
-  column_view const& values,
-  size_type offset,
-  scalar const& fill_value,
+std::pair<std::unique_ptr<table>, std::unique_ptr<table>> groupby::shift(
+  table_view const& values,
+  host_span<size_type const> offsets,
+  std::vector<std::reference_wrapper<const scalar>> const& fill_values,
   rmm::mr::device_memory_resource* mr)
 {
   CUDF_FUNC_RANGE();
-  CUDF_EXPECTS(values.type() == fill_value.type(),
-               "values and fill_value should have the same type.");
-
-  auto stream         = rmm::cuda_stream_default;
-  auto grouped_values = helper().grouped_values(values, stream);
+  CUDF_EXPECTS(values.num_columns() == static_cast<size_type>(fill_values.size()),
+               "Mismatch number of fill_values and columns.");
+  CUDF_EXPECTS(
+    std::all_of(
+      thrust::make_counting_iterator(0),
+      thrust::make_counting_iterator(values.num_columns()),
+      [&](auto i) { return values.column(i).type().id() == fill_values[i].get().type().id(); }),
```

Comment:
Does fill values take care of converting scale for decimal values?
because instead of checking `data_type==data_type`, type id is only checked here. it misses scale.
If not, cahnge to check type() equality instead of id.

Reply:
Good point, @karthikeyann . I'd prefer we determine fill values upstream and let this function throw if scales don't match. Easy enough to write a utility function to create "proper" fill values given loose types/columns. Putting that logic here wouldn't be re-usable.